boot/header.S

```diff
@@ -182,6 +218,166 @@ section_table:
 	.word	0				# NumberOfLineNumbers
 	.long	0x60500020			# Characteristics (section flags)
 
+boot_done:
+	# Print the message. Called with bh=0, and dh, dl already set.
+
+	movw	$boot_msg, %bp
+	movw	$(boot_msg_end - boot_msg), %cx
+	movb	$0x07, %bl			# Attribute 7 (normal).
+	movw	$0x1301, %ax			# Write string, move cursor.
+	int	$0x10
+
+load_setup:
+	xorw	%dx, %dx			# dl=drive 0, dh=head 0 for below.
+	xorb	%ah, %ah
+	int	$0x13				# reset FDC
+
+	# Load the setup sectors directly after the boot block.
+	# Note that 'es' is already set up.
+
+	movw	$0x0002, %cx			# sector 2, track 0
+	movw	$0x0200, %bx			# address = 512, in BOOT_SEG
+	movw	$(0x0200 + SETUP_SECS), %ax	# service 2, nr of sectors
+						# (assume all on head 0, track 0)
+	int	$0x13				# read it.
+	jc	load_setup			# error - repeat
+
+	xor	%bx, %bx
+	movb	$0x03, %ah			# Read cursor pos.
+	int	$0x10
+
+	movw	$header_msg, %bp
+	movw	$(header_msg_end - header_msg), %cx
+	movb	$0x07, %bl			# Attribute 7 (normal).
+	movw	$0x1301, %ax			# Write string, move cursor.
+	int	$0x10
+
+load_setup_done:
+	# Get disk drive parameters, specifically number of sectors/track.
+	# It seems that there is no BIOS call to get the number of sectors.
+	# Guess 18 sectors if sector 18 can be read, 15 if sector 15 can be
+	# read. Otherwise guess 9.
+
+	movw	$MAIN_SEG, %ax
+	movw	%ax, %es
+	xorw	%bx, %bx 		# bx is starting address within segment
+
+	xorw	%dx, %dx		# drive 0, head 0
+	movw	$0x0012, %cx		# sector 18, track 0
+	movw	$0x0201, %ax		# service 2, 1 sector
+	int	$0x13
+	jnc	got_sectors
+	movb	$0x0f, %cl		# sector 15
+	movw	$0x0201, %ax		# service 2, 1 sector
+	int	$0x13
+	jnc	got_sectors
+	movb	$0x09, %cl
+
+got_sectors:
+	movw	%cx, %ds:sectors
+
+	# Load the main test program.
+	call	read_it
+
+	# FDD motor will be switches off via floppy_off() called from
+	# global_init().
+
+	# Fix up the Linux boot header to indicate we've loaded into low memory.
+	movl	$LOW_LOAD_ADDR, code32_start
+
+	# After that (everything loaded), we jump to the setup code loaded
+	# directly after the boot block.
+
+	ljmp	$SETUP_SEG, $0
+
+# This subroutine loads the system at address 0x10000, making sure no 64KB
+# boundaries are crossed. We try to load it as fast as possible, loading
+# whole tracks whenever we can.
+#
+# in:	es - starting address segment (normally 0x1000)
+#
+
+read_it:
```

Comment:
I see a single reference to this label in this commit. Can this subroutine be inlined ? Although if it can, it remains to be seen whether it actually saves space, of course :)

Reply:
Yes, very god point, thanks for a flagging and thank you for a prompt initial review.

Indeed, I also noticed that and I think we can save 4 bytes there (3 from the call, one from ret). Maybe even more with some changes in the code flow and by using more registers. I also would like to reduce the number of variables kept in the memory.

As mentioned, there are several more opportunities, which I will definitely explore. Just wanted to show what I have been working on and get some initial feedback. If we decide it makes sense, I promise we can fit the proper project name there before I consider this done, so more changes will follow. ;)

Perhaps  we can even fit the functionality of mbr.S, but don't want to over-promise.